bxserum/connections/ws.go

```diff
@@ -5,84 +5,325 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"github.com/bloXroute-Labs/serum-client-go/utils"
 	"github.com/gorilla/websocket"
 	"github.com/sourcegraph/jsonrpc2"
 	"google.golang.org/protobuf/encoding/protojson"
 	"google.golang.org/protobuf/proto"
-	"strings"
+	"sync"
+	"time"
 )
 
-func WSRequest[T proto.Message](conn *websocket.Conn, request []byte, response T) error {
-	err := sendWS(conn, request)
+const (
+	handshakeTimeout       = 5 * time.Second
+	subscriptionBuffer     = 1000
+	unsubscribeGracePeriod = 3 * time.Second
+)
+
+type WS struct {
+	messageM      sync.Mutex
+	subscriptionM sync.RWMutex
+	requestID     *utils.RequestID
+	conn          *websocket.Conn
+	ctx           context.Context
+	cancel        context.CancelFunc
+	err           error
+	writeCh       chan []byte
+
+	requestMap      map[uint64]requestTracker
+	subscriptionMap map[string]subscriptionEntry
+}
+
+func NewWS(endpoint string) (*WS, error) {
+	dialer := websocket.Dialer{HandshakeTimeout: handshakeTimeout}
+	conn, _, err := dialer.Dial(endpoint, nil)
 	if err != nil {
-		return err
+		return nil, err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	ws := &WS{
+		requestID:       utils.NewRequestID(),
+		conn:            conn,
+		ctx:             ctx,
+		cancel:          cancel,
+		writeCh:         make(chan []byte, 100),
+		requestMap:      make(map[uint64]requestTracker),
+		subscriptionMap: make(map[string]subscriptionEntry),
+	}
+	go ws.readLoop()
+	go ws.writeLoop()
+	return ws, nil
+}
+
+func (w *WS) readLoop() {
+	defer w.cancel()
+
+	for {
+		// if lock is held by another processing routine, then wait for release before processing any further messages
+		// typically, messages can be dispatched independently, but in some cases it may be important to finish processing a single message before allowing the socket to read any more (e.g. subscription response before processing potential updates)
+		w.messageM.Lock()
+		w.messageM.Unlock()
```

Comment:
Not quite. All reading from the websocket is done on a single goroutine, and each message processing happens on a different one, so messages don't necessarily get processed in order. In most cases this is fine since we're just mapping request IDs to responses, but for subscriptions we're going to drop some initial subscription updates if the response isn't processed first (since the map entry won't be setup correctly).

In this implementation, `w.processRPCResponse(response)` may acquire the lock, which will prevent subsequent execution of this loop until it's released. The goroutine that actually handles the messages becomes responsible for indicating when it's done holding the message processing lock. I'll flesh out this comment a bit more.